_drafts/advanced-gdb.md

```diff
@@ -0,0 +1,993 @@
+
+```
+(gdb) set $wifi = mgmt_thread_data.next_thread.next_thread
+(gdb) p $wifi
+$11 = (struct k_thread *) 0x200022ac <eswifi_spi0+20>
+```
+
+You can also use convenience variables to help you print fields within an array of structs.
+
+```
+(gdb) p &shell_wifi_commands
+$37 = (const struct shell_static_entry (*)[5]) 0x8031650 <shell_wifi_commands>
+```
+
+The above array is a list of `shell_static_entry` structs, each of which has many fields. Browsing through lists of structs is sometimes cumbersome, especially if I'm only trying to look at a single field.
+
+Convenience variables can help with this.
+
+Let's print the `help` element from struct in the array. I'll set `$i = 0` as my index counter and use `$i++` in each command so that it increments.
+
+```
+(gdb) set $i = 0
+(gdb) print shell_wifi_commands[$i++]->help
+$1 = 0x80314e0 "\"<SSID>\"\n<SSID length>\n<channel number (optional), 0 means all>\n<PSK (optional: valid only for secured SSIDs)>"
+(gdb) <enter>
+$2 = 0x803155c "Disconnect from Wifi AP"
+...
+```
+
+Each time I press enter, the previous command is executed and `i` increments, printing me the data in the next struct in the array.
+
+[Reference](https://sourceware.org/gdb/onlinedocs/gdb/Convenience-Vars.html)
+
+{:.no_toc}
+
+### The `$` Variable
+
+In GDB, the value of `$` is the value returned from the previous command.
+
+```
+(gdb) p "hello"
+$63 = "hello"
+(gdb) p $
+$64 = "hello"
+```
+
+This is useful in its own right, but it is especially useful for helping me with one of my least favorite tasks in GDB: iterating over linked lists.
+
+You may have done something similar to the following:
+
+```
+(gdb) p mgmt_thread_data.next_thread
+$65 = (struct k_thread *) 0x2000188c <eswifi0+48>
+(gdb) p mgmt_thread_data.next_thread.next_thread
+$66 = (struct k_thread *) 0x200022ac <eswifi_spi0+20>
+```
+
+Instead of adding `.next_thread` onto the end of the list each iteration, you can use the value of `$` and just keep pressing `<enter>`.
+
+```
+(gdb) p mgmt_thread_data.next_thread
+$67 = (struct k_thread *) 0x2000188c <eswifi0+48>
+(gdb) p $.next_thread
+$68 = (struct k_thread *) 0x200022ac <eswifi_spi0+20>
+(gdb) <enter>
+$69 = (struct k_thread *) 0x2000a120 <k_sys_work_q+20>
+```
+
+[Reference](https://sourceware.org/gdb/onlinedocs/gdb/Value-History.html#Value-History)
+
+## Artificial Arrays
+
+It is often useful to print a contiguous region of memory as if it were an array. A common occurrence of this is when using `malloc` to allocate a buffer for a list of integers.
+
+```c
+int num_elements = 100;
+int *elements = malloc(num_elements * sizeof(int));
+```
+
+In GDB, if you try to print this, it will just print the pointer value, since it doesn't know it's an array.
+
+```
+(gdb) p num_elements
+$1 = 100
+(gdb) p elements
+$2 = (int *) 0x5575e51f6260
+```
+
+We can print this entire array using one of two ways. First, we can cast it to a `int[100]` array and print that.
+
+```
+(gdb) p (int[100])*elements
```

Comment:
interesting, i've always written as `*(int[100] *)elements` but I guess the extra * is not necessary